aten/src/ATen/cuda/CUDAGraph.cpp

```diff
@@ -315,4 +338,261 @@ CUDAGraph::~CUDAGraph() {
+                .dstPos = memcpyParams1.dstPos,
+                .dstPtr = dstPtr,
+                .extent = memcpyParams1.extent,
+                .kind = memcpyParams1.kind
+              },
+            },
+          };
+        },
+      });
+    } else if (type == cudaGraphNodeTypeMemset) {
+      cudaMemsetParams memsetParams1;
+      AT_CUDA_CHECK(cudaGraphMemsetNodeGetParams(node, &memsetParams1));
+
+      auto dstPtrResult = checkAllocationWithinGraph(memsetParams1.dst, allocations_);
+      if (!dstPtrResult) {
+        continue;
+      }
+      
+      graph_node_param_updates_.push_back({
+        .node = node,
+        .compute_new_params = [memsetParams1, dstPtrResult](std::vector<void*> actualDataPtrs) {
+          void* dstPtr = memsetParams1.dst;
+          if (dstPtrResult) {
+            auto [alloc_idx, offset] = *dstPtrResult;
+            dstPtr = (char*)actualDataPtrs[alloc_idx] + offset;
+          }
+          return cudaGraphNodeParams{
+            .type = cudaGraphNodeTypeMemset,
+            .memset = {
+              .dst = dstPtr,
+              .pitch = memsetParams1.pitch,
+              .value = memsetParams1.value,
+              .elementSize = memsetParams1.elementSize,
+              .width = memsetParams1.width,
+              .height = memsetParams1.height,
+            },
+          };
+        },
+      });
+    } else if (type == cudaGraphNodeTypeGraph || type == cudaGraphNodeTypeConditional) {
+      throw std::runtime_error("horrible. please don't make me support these.");
+      assert(false);
+    } else if (
+      type == cudaGraphNodeTypeEmpty ||
+      type == cudaGraphNodeTypeWaitEvent ||
+      type == cudaGraphNodeTypeHost ||
+      type == cudaGraphNodeTypeEventRecord ||
+      type == cudaGraphNodeTypeExtSemaphoreSignal ||
+      type == cudaGraphNodeTypeExtSemaphoreWait ||
+      type == cudaGraphNodeTypeMemAlloc ||
+      type == cudaGraphNodeTypeMemFree
+    ) {
+      // this is fine
+    } else {
+      throw std::runtime_error("graph node type unknown");
+      assert(false);
+    }
+  }
+}
+
+void CUDART_CB hostMemoryFreeCallback(cudaStream_t stream, cudaError_t error, void* data) {
+  at::getCPUAllocator()->raw_deallocate(data);
+}
+
+void CUDAGraph::replay_dynamic(std::vector<void*> prefilledDataPtrs, std::vector<size_t> prefilledLens) {
+  TORCH_CHECK(num_dynamic_args_, "Must be a dynamic graph");
+  TORCH_CHECK(has_graph_exec_,
+              "Called CUDAGraph::replay_dynamic without a preceding successful capture.");
+  TORCH_CHECK(prefilledDataPtrs.size() == (size_t) num_dynamic_args_);
+  TORCH_CHECK(prefilledLens.size() == prefilledDataPtrs.size());
+  c10::OptionalDeviceGuard device_guard{capture_stream_.device()};
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+
+  // take the allocations that were requested during the capture, and allocate them "for real"
+  std::vector<void*> actualDataPtrs;
+  auto allocator = c10::cuda::CUDACachingAllocator::get();
```

Comment:
How might this interact with MemPool's concept of a current allocator?

Reply:
By the time it gets here, the capture is done. So we are not messing with the allocator at this point. So this is just a normal standard allocation. If this is not the standard way to get some quick scratch space on the GPU then I'll change it, but, I took a quick look around aten and this looked fairly common